tripal/tripal
# Gff3 performance and unit tests

# Bug Fix

Issue #1040 

## Description



 - Use cvterm lookup table in helper methods to increase performance. Previously, we were using a lookup table for cvterms within the main `loadGFF3` method, but we were not using the table in the helper methods, which do most of the actual inserting to the database. I have added a private method called `getCvterm` which first checks the lookup table before starting a db query. This has improved load time on a 1000 line gff file from about 23 seconds to about 11 seconds. More optimizations may be necessary in the future, but this seems like a good improvement for now.
 - Remove "use transaction" option from GFF Importer. All Tripal Importers are currently wrapped in a transaction, and the GFF Importer itself never begins a transaction. This means that the "use transaction" option does not actually have any functional purpose, and is therefore a bit confusing and misleading, so I have removed it.
 - Added unit tests for GFF Importer Attributes. The tests I have added do not test every single possible case, but they test for most of the main GFF attributes that are coded for in the GFF Importer, including the feature, alias, dbxref, ontology term, parent, target, properties, and derives from relationship.
 - Added small .gff and .fasta files to be used for attribute unit tests.

## Testing?





- It may be necessary to clear your cache after pulling this code, which can be done via `drush cc all`.
- Load your .fasta file with the Tripal FASTA importer
- Navigate to `admin/tripal/loaders/chado_gff3_loader`
- Fill out the form and submit your GFF Loader job
- Verify that the loader works and completes faster than before.
- If you would like to run the unit tests locally:
  - Navigate to the tripal module directory
  - Install dependencies with `composer install`
  - Run all unit tests (`./vendor/bin/phpunit`) or run only the new unit tests added in this PR with
```
./vendor/bin/phpunit --filter testGFFImporterAttribute tests/tripal_chado/loaders/GFF3ImporterTest.php
```
